libmscore/instrument.cpp

```diff
@@ -804,6 +804,15 @@ void Channel::switchExpressive(MasterSynthesizer* m, bool expressive, bool force
       if ((_userBankController && !force) || !m)
             return;
 
+      // Check that we're actually changing the MuseScore General soundfont
+      if (!force) {
+            SynthesizerGroup fluid = m->state().group("Fluid");
+            if (!fluid.front().data.contains("MuseScore_General")) {
```

Comment:
Why is it needed to limit this function to work only with `MuseScore_General` soundfont? And couldn't it happen to break because of some file/soundfont renaming or something like that?

Reply:
We already have a lot of bits that would break if the soundfont were renamed. Ideally we'd have a global constant for the soundfont filename, but I'll leave that for another PR. This function's only needed for MuseScore general because MuseScore general has two versions of each patch, a normal one and a single note dynamics enabled one. We need to switch to the single note dynamics enabled one in this function, but since other soundfonts don't have the same organisation, we could up ending doing weird things if we ran this function with other soundfonts.

Reply:
Are there another options to distinguish between soundfonts? Maybe some metadata contained in the soundfont file or something like that? Using a file name to do it sound very easy to break, and I would rather avoid using this way of soundfont detection if it is possible.

As far as I understand, the old way was to check whether the patch with the given bank and program number has "Expr." in its name, wasn't it? What are the reasons to replace it? At first glance it sounds a not very clean but still better way since it relies on the soundfont structure rather than a file name.

Reply:
The problem with that was it worked one way - going from non expressive to expressive - but it didn't work the other way, because I was removing the expr from the patch name, and then checking to see if that was in the non expressive patch name. This didn't work sometimes - going from Cello Expr to Cello was fine, but from Hm Mute Expr to Harmon  Mute didn't work at all. It was too unreliable.

AFAIK at the moment there's no way of checking the metadata of loaded soundfonts, not with out current interface at least. The name is the most we can get out of it.

Reply:
Woudn't it be enough to check for "Expr." substring presence (maybe at certain position) in the patches that are expected to be expressive? Then such names variations wouldn't be a problem.

Reply:
Isn't that just as prone to problems as using the soundfont name? What if one soundfont has, I don't know, an 'Expressophone' patch? Or any thing else with Expr in it? It's just as, if not more prone to problems than using the soundfont name...

Reply:
Also, results for searching for MuseScore_General: 28 results in 7 files

I think it's safer to use the file name than it is to check for Expr. Would you agree?

Reply:
> Also, results for searching for MuseScore_General: 28 results in 7 files

As far as I can see, these are mostly in README and CMakeLists.txt files where they actually should be. Also there are some entries concerning the definition of the default soundfont (like [this one](https://github.com/musescore/MuseScore/blob/00b8dc4012b4760e469e8a6c148d7e302a4563ac/libmscore/synthesizerstate.cpp#L84)) where it is also OK to use the file name.

I agree that using "Expr." substring search is also not fully reliable (I am very curious how an Expressophone could sound...), and I agree that making a restriction by soundfont name should be OK here but I believe it would be preferable to use the name recorded to the soundfont itself. The problem is that we indeed do not currently read it at all. It is not difficult to add this (see [this commit](https://github.com/dmitrio95/MuseScore/commit/3b96d830eafac0b1c0982637928b474f44378a8e)) but this still requires some changes in the synthesizer subsystem interface (they can be limited just to adding things though). Maybe that commit can be used as a foundation of the correct soundfont name usage.

By the way, the soundfont version is not taken into account with the file name approach, that could be a problem if using an older version of `MuseScore_General` soundfont (is this the case for anyone?) The soundfont name contained in `MuseScore_General` soundfonts themselves happens to contain the version information so this can be used to overcome also this issue.

So could this check be switched to using the soundfont name/version data from the soundfont itself, or do you believe that this is not so urgent or necessary to do?

Reply:
Thanks for your detailed response. I think I'll take your commit, and then build upon that, that should be enough to resolve this issue :)

Reply:
I seem to have committed something strange, [this](https://github.com/musescore/MuseScore/commit/610010d2fe162d9c29c89cfcb03e69ab264da1d5) should be closer to what it may be.

Also, for demonstration purposes, there are a couple of `qDebug()` messages there so that commit should probably anyway not be used just as it is.

Reply:
Right, should be fixed. I hope this PR can be merged now :)

Reply:
OK, thanks!